[docs] Add reference doc for `create-expo-app` (#29488)
expo/expo: docs/pages/more/create-expo.mdx

<Terminal cmd={['$ bun create expo']} />

</Tab>

</Tabs>

Running the above command will prompt you to enter the app name of your project. This app name is also used in the app config's [`name`](/versions/latest/config/app/#name) property.

<Terminal cmd={['What is your app named? my-app']} />

## Options

Uses the following options to customize the command behavior.

### `--yes`

Uses the default options to create a new project.

### `--no-install`

Skips installing npm dependencies or CocoaPods.

### `--template`

Running `create-expo-app` with a [Node Package Manager](#node-package-managers-support) initializes and sets up a new Expo project using the default template.

You can use the `--template` option to choose one of the following templates:

| Template                | Description                                                                                                                                                                           |
| ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Default                 | Default template. Designed to build multi-screen apps. Includes recommended tools such as Expo CLI, Expo Router library and TypeScript configuration enabled. Suitable for most apps. |

Reviewer: We can consider adding links to source of templates that we keep on GH. It could be a separate column, or we can link the name:
* https://github.com/expo/expo/tree/main/templates

```yml .yarnrc.yml
nodeLinker: node-modules
```

#### EAS installation

Yarn Modern on EAS requires an [additional configuration step](#additional-configuration-steps) and adding [`eas-build-pre-install` hook](/build-reference/npm-hooks/) in your project's **package.json**, add the following configuration:

```json package.json
{
  "scripts": {
    "eas-build-pre-install": "corepack enable && yarn set version 4"
  }
}
```

### pnpm

#### Local installation

Requires installing Node.js. See [pnpm documentation](https://pnpm.io/installation) for installation instructions.

#### Additional configuration steps

After installing pnpm on your local development environment, you've to specify the type linker to install npm packages using [`node-linker`](https://pnpm.io/npmrc#node-linker).

Create **.npmrc** in your project's root directory and add the following configuration:

```ini .npmrc
node-linker=hoisted
```

#### EAS installation

Supported by default if the project directory contains **pnpm-lock.yaml** and the [additional configuration step](#additional-configuration-steps-1) followed.

### Bun

> > **warning** Bun version `1.0.22` or below is supported due to [peer dependency conflicts](https://github.com/oven-sh/bun/issues/9135).

Reviewer: ```suggestion
> **warning** Only Bun versions `1.0.22` or below are supported due to [peer dependency conflicts](https://github.com/oven-sh/bun/issues/9135).
```

Let's clear the message if the "below" is not a typo, otherwise let's correct it.

Author: I'm going to remove this warning based on this message earlier this week: https://exponent-internal.slack.com/archives/C1QLM27T9/p1718034956190959?thread_ts=1718032071.445799&cid=C1QLM27T9

Reviewer: That makes sense, thanks for the context! 👍

### `--example`

Use this option to initialize a project using an example from [expo/examples](https://github.com/expo/examples).

For example, running `npx create-expo-app --example with-router` will set up a project with Expo Router library.

### `--version`

Prints the version number and exits.

### `--help`

Prints the list of available options and exits.

## Node Package Managers support

Creating a new project with `create-expo-app` also handles setting up additional configuration needed for a specific Node Package Manager.

**If you are migrating from one package manager to another**, you've to manually carry out the additional configuration in your project. **If you are using [EAS](/eas/)**, you also have to configure your project for any additional required steps manually.

All the additional steps for each package manager are listed below.

### npm

#### Local installation

npm is installed as part of Node.js installation. See [Node.js documentation](https://nodejs.org/en/download/package-manager) for installation instructions.

#### EAS installation

Supported by default if the project directory contains **package-lock.json**.

> ### Yarn Classic (v1)

Author: ```suggestion
### Yarn 1 (Classic)
```

#### Local installation

> Yarn Classic (v1) is usually installed as a global dependency of npm. See [Yarn Classic documentation](https://classic.yarnpkg.com/en/docs/getting-started) for installation instructions.

Author: ```suggestion
Yarn 1 (Classic) is usually installed as a global dependency of npm. See [Yarn 1 documentation](https://classic.yarnpkg.com/en/docs/getting-started) for installation instructions.
```

#### EAS installation

Supported by default if the project directory contains **yarn.lock**.

> ### Yarn Modern (v2+)

Author: ```suggestion
### Yarn 2+ (Modern)
```

#### Local installation

> Yarn Modern handles package management differently than Yarn Classic.
> One of the core changes in Yarn Modern is the PnP node linking model which does not work with React Native.

Author: ```suggestion
Yarn 2+ handles package management differently than Yarn 1. One of the core changes in Yarn 2+ is the PnP node linking model that does not work with React Native.
```

| ----------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [Default](https://github.com/expo/expo/tree/main/templates/expo-template-default)                     | Default template. Designed to build multi-screen apps. Includes recommended tools such as Expo CLI, Expo Router library and TypeScript configuration enabled. Suitable for most apps. |
| [Blank](https://github.com/expo/expo/tree/main/templates/expo-template-blank)                         | Installs minimum required npm dependencies without configuring navigation.                                                                                                            |
| [Blank (TypeScript)](https://github.com/expo/expo/tree/main/templates/expo-template-blank-typescript) | A Blank template with TypeScript enabled.                                                                                                                                             |

> | [ Navigation (TypeScript)](https://github.com/expo/expo/tree/main/templates/expo-template-tabs)       | Installs and configures file-based routing with TypeScript enabled.                                                                                                                   |

Reviewer: ```suggestion
| [Navigation (TypeScript)](https://github.com/expo/expo/tree/main/templates/expo-template-tabs)        | Installs and configures file-based routing with TypeScript enabled.                                                                                                                   |
```

Whitespace nit.

Author: Interesting. Prettier config didn't catch this on my system 😄
